lib/async_hooks.js

```diff
@@ -204,11 +206,142 @@ class AsyncResource {
   }
 }
 
+let activeAsyncLocalStorageCount = 0;
+const kLocalStorageList = Symbol('kLocalStorageList');
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const triggerResource = executionAsyncResource();
+    // Value of triggerResource is always a non null object
+    const localStorageList = triggerResource[kLocalStorageList];
+    if (!localStorageList) {
+      return;
+    }
+    resource[kLocalStorageList] = [];
```

Comment:
Done in https://github.com/nodejs/node/pull/26540/commits/1a9f461303f57a9ef54851123f161dfb03bd7cbc is that what you had in mind?

Reply:
Sorry for disturbing the conversation, but https://github.com/nodejs/node/commit/1a9f461303f57a9ef54851123f161dfb03bd7cbc still has the same problem, i.e. litter arrays generated on the hot path (global async hook). The thing is that is common scenario (when there is an enabled `AsyncLocalStorage`) `_propate()` method still allocates an array and pushes items into it. See line https://github.com/nodejs/node/pull/26540/files#diff-0bb01a51b135a5f68d93540808bac801R270

Copy on write approach for array propagation should fix the problem.

Reply:
Do you mean something like 
```js
resource[kLocalStorageList] = currentResource[kLocalStorageList]
```
before the `for(;;)` in `init` ? Actually that is probably enough here

Reply:
Yes, that's what I mean. Another part of the puzzle will be whether the array should be global (i.e. all resources will have reference to the same array) or it should have copy on write semantics (i.e. a copy of the array should be made when a storage is enabled or disabled). Most likely it should be copy on write option, but edge cases should be considered.

Reply:
propagating the array will be enough here I believe.

Reply:
> propagating the array will be enough here I believe.

It might be. BTW isn't it in fact the same as having a global `locals` array that was there previously?

Reply:
Not totally, the array is local to certain branches of the async tree. An instance can only be in one array, but multiple arrays can co-exist in the same process.

Reply:
@mcollina current behavior should be the fastest path I believe

Reply:
> Not totally, the array is local to certain branches of the async tree. An instance can only be in one array, but multiple arrays can co-exist in the same process.

Yes, and that might be a problem in combination with `.enable()`/`.disable()` methods. Imagine enabling/disabling an `AsyncLocalStorage` in one branch while it's actively used in another one. Or there might be a situation when there was an `AsyncLocalStorage` and a new instance is created and enabled/disabled in one of sub-branches. That new instance may "travel" into another sub-branch, as the same array is used instead of copy-on-write behavior.

Reply:
I moved back to the state with the checks in `_propagate` as it seems to be the faster path here.